Add Tektronix MSO Scope Series

OptimisticBeliever opened the pull request:
Two child classes have been tested with devices MSO58 and MSO64.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added support for Tektronix MSO58 and MSO64 oscilloscopes with waveform acquisition, trigger configuration, measurement capabilities, and image download functionality.

* **Documentation**
  * Added documentation for Tektronix MSO Series oscilloscope instruments.

* **Tests**
  * Added comprehensive test coverage for oscilloscope operations and device integration.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

OptimisticBeliever committed:
Add the tektronix Mso Series
Add the test with and without device

OptimisticBeliever committed:
reformat to avoid line too long (<100)

OptimisticBeliever committed:
reformat to avoid line too long (<100)

OptimisticBeliever committed:
Remove includeSCPI=True in the two child class
Add the measurement_gating properties

OptimisticBeliever committed:
Add the test with device for the MSO64

OptimisticBeliever committed:
Merge branch 'pymeasure:master' into dev/tektronixMSO

BenediktBurger commented:
Thanks for your contribution.
I had already a quick look a few days ago, but it is a large PR and will take some time.

A few quick points:
- Are these instruments similar to the Keithley ones? I read, that they fusioned in some point in history.
  If you know, that there are similarities or similar devices with both labels (maybe not yet in our codebase), you could link them in the documentation, similar to [HP devices](https://pymeasure.readthedocs.io/en/latest/api/instruments/hp/index.html) which link to Agilent/Keysight.
- Please add the API documentation: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#adding-documentation

OptimisticBeliever committed:
Add the API documentation for the tektronix MSO Series oscilloscopes.

OptimisticBeliever committed:
Update docstring

OptimisticBeliever committed:
Merge remote-tracking branch 'origin/dev/tektronixMSO' into dev/tektronixMSO

OptimisticBeliever committed:
Include tektronixMsoSeries.rst into index.rst

BenediktBurger commented on docs/api/instruments/tektronix/tektronixMsoSeries.rst:
As you add this manufacturer for the first time, you have to add this file to the index in the parent directory as well.

OptimisticBeliever committed:
Correct typos

OptimisticBeliever committed:
Correct Docs and Linting

OptimisticBeliever committed:
Correct Docs and Linting

OptimisticBeliever committed:
Change copyright year.

OptimisticBeliever committed:
Merge branch 'pymeasure:master' into dev/tektronixMSO

OptimisticBeliever committed:
Add position property to class TektronixMsoScopeChannel

BenediktBurger commented:
By the way, you do not have to cover all lines with tests.

OptimisticBeliever committed:
Add horizontal_mode, horizontal_manualmode_cfg, horizontal_record_length, horizontal_sample_rate, horizontal_delay_mode, horizontal_minsamplerate_override, horizontal_minsamplerate_value, horizontal_delay_time properties to class TektronixMsoScope

OptimisticBeliever committed:
Add test_horizontal_minsamplerate

OptimisticBeliever committed:
Add method measurement_clear_all

OptimisticBeliever committed:
Add class TektronixMsoScopeMathChannel
Add methods add_new_math and math_delete_all to class TektronixMsoScope

OptimisticBeliever committed:
Add testMath_basic
Add test_math_fft

OptimisticBeliever committed:
Docs and linting errors

OptimisticBeliever committed:
Add the TektronixMsoScopeMathChannel to the tektronixMsoSeries.rst

OptimisticBeliever commented:
> Thanks for your contribution. I had already a quick look a few days ago, but it is a large PR and will take some time.
> 
> A few quick points:
> 
> * Are these instruments similar to the Keithley ones? I read, that they fusioned in some point in history.
>   If you know, that there are similarities or similar devices with both labels (maybe not yet in our codebase), you could link them in the documentation, similar to [HP devices](https://pymeasure.readthedocs.io/en/latest/api/instruments/hp/index.html) which link to Agilent/Keysight.
> * Please add the API documentation: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#adding-documentation

Hello Benedikt,
no this device is not linked to keithley instruments.
Anyway could you please what is wrong in the Docs and Linting auto check. I could not figure it out !

Thanks,

BenediktBurger reviewed (commented):
I'm not sure, what causes the issue, but marked two points, where it could be.

You can check it on your computer, if you run `make html` in a console (not python console) in the docs directory.

OptimisticBeliever committed:
Update pymeasure/instruments/tektronix/tektronixMsoSeries.py

Co-authored-by: Benedikt Burger <67148916+BenediktBurger@users.noreply.github.com>

OptimisticBeliever commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1344 @@
+        - "label": "" (str)
+        - "volts_div": vertical divisions (float)
+        - "trigger_level": trigger level (float)
+        """
```
Hello Benedikt, seems to be there. I don't have sphinx-build installed on my environment so I could not test it as you suggested. Anyway there is no more auto-checks failure. Could we close this conversation ?

BenediktBurger replied:
I tried to hint at possible positions.

I'm glad, that the docs builds succeeded (on CI).

OptimisticBeliever replied:
Many thanks. Have a nice day.

OptimisticBeliever committed:
Change the discrete value "NORM" to "NORMAL" for trigger_mode property

OptimisticBeliever committed:
Typo error in the set_command of trigger_edge_coupling property

OptimisticBeliever committed:
remove map values=True for the trigger_width_polarity property

OptimisticBeliever committed:
change the values from ["NEGATIVE", "POSITIVE"] to {"negative": "NEGATIVE", "positive": "POSITIVE"} for the trigger_width_polarity property

OptimisticBeliever committed:
change the values from {.."lowpass": "HFREJ", "highpass": "LFREJ"..} to {.."lowpass": "LFREJ", "highpass": "HFREJ"..} for the _TRIGGER_COUPLING dictionary

OptimisticBeliever committed:
change the _TRIGGER_TYPES to a dictionary

OptimisticBeliever committed:
change the TRIGGER_TYPES values

OptimisticBeliever committed:
Change the ANALOG_TRIGGER_SOURCE to a dictionary
Update the trigger_edge_source and trigger_width_source properties

OptimisticBeliever committed:
Update the TRIGGER_SOURCE
Add the test_trigger_width_source
Update the test_measurement_configure

OptimisticBeliever committed:
change the values of the bwlimit property to a dictionary

OptimisticBeliever committed:
change the BANDWIDTH_LIMITS values

OptimisticBeliever committed:
change the bandwidth_limit in the test_ch_current_configuration

OptimisticBeliever committed:
linting error

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -851,7 +851,8 @@ def center_trigger(self):
         """,
         validator=strict_discrete_set,
-        values=["NEGATIVE", "POSITIVE"],
+        values={"negative": "NEGATIVE", "positive": "POSITIVE"},
+        map_values=True,
```
Why not:
```suggestion
        values={False: "NEGATIVE", True: "POSITIVE"},
        map_values=True,
```
maybe also renaming the control to `..positive_polarity`.

Just as a suggestion.

OptimisticBeliever replied:
Hello Benedikt. For this properties relative to slope and polarity I want to keep it this way. The reason is that Lecroy Scopes has those properties merged i.e slope and polarity. This is not the case for Tektronix scopes. I had an abstraction layer that use Pymeasure to rule any scope and this way I can have negative/positive for both properties whatever the scope used.

BenediktBurger replied:
Ok

OptimisticBeliever committed:
Change _measurable_parameters list to a dictionary

OptimisticBeliever committed:
Add map_values=True to the measurement_add property

OptimisticBeliever committed:
Update the method measurement_configure

OptimisticBeliever committed:
Change the BANDWIDTH_LIMITS float to string list
Change the MEAS_SLOTS list values
Change the EXPECTED_MEAS_VALUES keys
Add the MEASURABLE_PARAMETERS

Add the test_measurement_add

OptimisticBeliever committed:
Add saving screenshot to computer drive on test_download_image

OptimisticBeliever committed:
Add *OPC? in method download_image
Add the break_on_termchar=True in method download_image

OptimisticBeliever committed:
Add clear method

OptimisticBeliever committed:
Add test_math_define

OptimisticBeliever committed:
Update the ANALOG_TRIGGER_SOURCE in class TektronixMSO64

OptimisticBeliever committed:
Add _grid_number, _header_size, _footer_size, waveform_source attributes
Add a validator and values to the horizontal_record_length property
Add waveform_data_source, waveform_points, waveform_first_point, waveform_last_point, waveform_preamble properties
Add _fill_yaxis_preamble, _digitize, _header_footer_sanity_checks, _npoints_sanity_checks, _acquire_data, _process_data, download_waveform methods

OptimisticBeliever committed:
Add test_waveform_preamble, test_download_data, test_download_data_all_points

OptimisticBeliever committed:
Merge branch 'pymeasure:master' into dev/tektronixMSO

OptimisticBeliever committed:
correct Docs and linting

OptimisticBeliever committed:
remove trailing whitespace

OptimisticBeliever committed:
remove includeSCPI=True
set map_values=False for bwlimit property

OptimisticBeliever committed:
add inheritance of the `SCPIMixin` class

OptimisticBeliever committed:
change BANDWIDTH_LIMITS = [[20.0000E+6, "20MHz"], [250.0000E+6, "250MHz"], [1.0000E+9, "1GHz"]]

OptimisticBeliever committed:
instr.ch_1.bwlimit = "20MHz"

OptimisticBeliever committed:
Add test_trigger_lower_threshold, test_trigger_upper_threshold

OptimisticBeliever committed:
Replace str to float in test_bwl

OptimisticBeliever committed:
Merge branch 'pymeasure:master' into dev/tektronixMSO

BenediktBurger commented:
Hey @OptimisticBeliever , there is another PR regarding an additional instrument (#1172) for this series.
When do you expect this PR to be ready to be merged?

OptimisticBeliever committed:
Merge branch 'pymeasure:master' into dev/tektronixMSO

Commit:
Set bwlimit validator as strict_discrete_set

Commit:
BANDWIDTH_LIMITS set to simple list of strings

Commit:
remove WAI and add OPC in download_image

Commit:
add try except  with WAI

Commit:
Add the deskew property

OptimisticBeliever committed:
Merge branch 'pymeasure:master' into dev/tektronixMSO

Commit:
Add 'deskew': 0.0 to the test_ch_current_configuration

Commit:
Merge remote-tracking branch 'origin/dev/tektronixMSO' into dev/tektronixMSO

Commit:
Add deskew to test_channel_setup

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2024 PyMeasure Developers
```
```suggestion
# Copyright (c) 2013-2026 PyMeasure Developers
```

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    truncated_range, strict_discrete_range
+
+
+def sanitize_source(source):
+    """Parse source string.
```
The docstring contradicts the name

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    coupling = Instrument.control(
+        "CH{ch}:COUPling?",
+        "CH{ch}:COUPling %s",
+        """Control the input coupling setting""",
```
Maybe add the three options to the docstring, especially as I do not understand what dcrej means (but I don't have the instrument).

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    deskew = Instrument.control(
+        "CH{ch}:DESKew?",
+        "CH{ch}:DESKew %G",
+        """Control the horizontal deskew time for the specified channel.
```
In which units?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    offset = Instrument.control(
+        "CH{ch}:OFFSet?",
+        "CH{ch}:OFFSet %G",
+        """Control the center of the screen in Volts by a a float parameter.
+        The range of legal values varies depending on range and scale. If the specified
+        value is outside of the legal range, the offset value is automatically set to the nearest
+        legal value.
```
Typically, we add the type (float) in the end, see: https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings

e.g. "(float)" at the end.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    position = Instrument.control(
+        "CH{ch}:POSition?",
+        "CH{ch}:POSition %d",
+        """Control the vertical position.""",
```
In which unit?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        """Control the vertical position.""",
+    )
+
+    display = Instrument.control(
+        "DISplay:GLObal:CH{ch}:STATE?",
+        "DISplay:GLObal:CH{ch}:STATE %d",
+        """Control the global state (display mode On or Off)""",
+        validator=strict_discrete_set,
+        values={True: 1, False: 0},
+        map_values=True,
```
For boolean properties, we suggest names like `display_enabled` to make sure what True/False means, e.g. `display_enabled=True`.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    probe_ext_attenuation = Instrument.control(
+        "CH{ch}:PROBEFunc:EXTAtten?",
+        "CH{ch}:PROBEFunc:EXTAtten %s",
+        """Control the attenuation value as a multiplier to the given scale factor.
+        """,
```
Here you could add the the values: 
```suggestion
        """Control the attenuation value as a multiplier to the given scale factor (int strictly between 1 and 10).
        """,
```

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    trigger_lower_threshold = Instrument.control(
+        "TRIGger:A:LOWerthreshold:CH{ch}?",
+        "TRIGger:A:LOWerthreshold:CH{ch} %f",
+        """Control the A lower trigger level threshold.
```
What is the "A" lower trigger threshold?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    math_type = Instrument.control(
+        "MATH:MATH{ch}:TYPe?",
+        "MATH:MATH{ch}:TYPe %s",
+        """Control the math type.""",
```
Add the possible values to the docstring.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        values=['CH1', 'CH2', 'CH3', 'CH4', 'CH5', 'CH6', 'CH7', 'CH8'],
+    )
+
+    math_average_mode = Instrument.control(
+        "MATH:MATH{ch}:AVG:MODE?",
+        "MATH:MATH{ch}:AVG:MODE %d",
+        """Control the math average mode flag.""",
+        validator=strict_discrete_set,
+        values={True: "ON", False: "OFF"},
+    )
```
Also, I propose to rename it `math_average_mode_enabled`, such that it is self-evident what True/False mean.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        """Add a new math channel."""
+        self.write(f'MATH:ADDNEW \"MATH{slot}\"')
+
+    math_define = Instrument.control(
```
As the whole class is a Math class, you do not need to prefix all properties with `math`.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+                "KAISERBESSEL", "GAUSSIAN", "FLATTOP2", "TEKEXPONENTIAL"],
+    )
+
+    math_FFT_horizontal_unit = Instrument.control(
```
Again, consider using a boolean property like `math_FFT_horizontal_log_enabled` which makes it easier to use (True/False) instead of strings like (LOG/LINEAR) which are always error prone.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        values=["LOG", "LINEAR"],
+    )
+
+    math_FFT_vertical_unit = Instrument.control(
```
here as well I'd use dbm enabled.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        values=["LINEAR", "DBM"],
+    )
+
+    math_FFT_view_autoscale = Instrument.control(
```
Also here with `autoscale_enabled`.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        depending on the commands and connection latency.
+    """
+
+    _BOOLS = {True: "ON", False: "OFF"}
```
As you use this dict also in other classes, I suggest to move it to the top level of the file, such that everyone could use it.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+                 name="Tektronix Oscilloscope",
+                 analog_channels=8,
+                 **kwargs):
+        super().__init__(adapter, name=name, **kwargs)
```
You could (and should) give the timeout to the super call instead of doing it in the next step manually.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+
+        # noinspection PyAttributeOutsideInit
+        if not hasattr(self, "channels"):
+            self.channels = {}
```
why don't you create that empty dict in the init call?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+
+        self.update_channels(number_of_channels=number_of_channels)
+
+    def update_channels(self, number_of_channels=None, **kwargs):
```
Do you need this method dynamically or only once at startup (i.e. can the number of channels change?)

The idea is that the class autodetects which device it is and how many channels it has, right?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+            return getattr(self, f"ch_{source if isinstance(source, int) else source[-1]}")
+
+    def autoscale(self):
+        """ sets the vertical, horizontal, and trigger controls
```
Docstrings should be in imperative mode and have a one line summary, so `Set`, not `sets`...

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        self.write("AUTOSet EXECute;*WAI")
+
+    def reset(self):
+        """ Resets the instrument. """
```
```suggestion
        """Reset the instrument."""
```

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        self.write("*RST;*WAI")
+
+    def clear(self):
+        """This command clears acquisitions, measurements, and waveforms."""
```
```suggestion
        """Clears acquisitions, measurements, and waveforms."""
```

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+            self._seconds_since_last_write = seconds_since_last_write
+        super().write(command, **kwargs)
+
+    _comm_header = Instrument.control(
```
Here again I recommend _comm_header_enabled.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    ##################
+    # Timebase Setup #
+    ##################
+    horizontal_mode = Instrument.control(
```
Also here, consider using horizontal_auto_mode_enabled as a name and then a boolean property.

Do it similarly for other two-option properties.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        "HORizontal:SAMPLERate:ANALYZemode:MINimum:VALue %s",
+        """Control the minimum sample rate used by Analysis Automatic horizontal mode.
+        Value is in sample per seconds.
+        AUTOmatic allows the instrument to set the minimum value.
```
Values contains `AUTOMATIC`, not `AUTO` or `AUTOmatic`.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        }
+        return tb_setup
+
+    def timebase_setup(self, scale=None, offset=None):
+        """Set up timebase. Unspecified parameters are not modified. Modifying a single parameter
+        might impact other parameters. Refer to oscilloscope documentation and make multiple
+        consecutive calls to timebase_setup if needed.
+
+        :param scale: interval in seconds between the trigger event and the reference position.
+        :param offset: horizontal scale per division in seconds/div.
+        """
+
+        if scale is not None:
+            self.timebase_scale = scale
+        if offset is not None:
+            self.timebase_offset = offset
```
Why don't you make it a setter for the `timebase` property?
In this case, only the getter's docstring is used and should start with `Control`.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    ###############
+    # Acquisition #
+    ###############
+    acquisition_state = Instrument.control(
```
acquisition_enabled would be better, wouldn't it?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    )
+
+    def run(self):
+        """Starts repetitive acquisitions.
```
```suggestion
        """Start repetitive acquisitions.
```

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        """Get the number of waveform points to be transferred with
+        the digitize method (int).
+
+        Note that the oscilloscope may provide less than the specified nb of points.
```
```suggestion
        Note that the oscilloscope may provide less than the specified number of points.
```

These few extra chars don't hurt but help understanding it.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        OUTside causes a trigger when a pulse is detected that is outside of a range set
+        by two values.
+        """,
+        validator=strict_discrete_set,
+        values=["LESSthan", "MOREthan", "EQual", "UNEQual", "WIThin", "OUTside"],
+        map_values=True,
+    )
```
That's not true: The list is mapped to integers.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+            _ = self.read()  # read '1'
+        except Exception:
+            # Fallback if *OPC? is not supported by your abstraction
+            self.write('*WAI')
```
Why should it happeN?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    )
+
+    def measurement_clear_all(self):
+        """deletes all the active instances of measurements defined in the scope application."""
```
```suggestion
        """Delete all the active instances of measurements defined in the scope application."""
```

Docstrinsgs should be imperative mode. Apply to all the others.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        """deletes all the active instances of measurements defined in the scope application."""
+        self.write("MEASUrement:DELETEALL")
+
+    def measurement_population_config(self, slot: int, global_flag=0,
+                                      limit_state=0, limit_value=1000):
+        """Configure the measurement population settings.
```
Why the mix of type hint and not type hint?

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+    def measurement_population_config(self, slot: int, global_flag=0,
+                                      limit_state=0, limit_value=1000):
+        """Configure the measurement population settings.
+        :param slot
+        :param global_flag : int
```
You did not complete the slot entry.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+        settings.
+        :param limit_state : int 0 turns off the population limit.
+        1 turns on the population limit.
+        :param limit_value : int population limit value for the measurement.
```
You should indent the lines not being a param.

BenediktBurger commented on pymeasure/instruments/tektronix/tektronixMsoSeries.py:
```diff
@@ -0,0 +1,1688 @@
+
+    def __init__(self, adapter,
+                 name="Tektronix MSO58 Oscilloscope",
+                 active_channels=8,
+                 **kwargs):
+        super().__init__(adapter, name, active_channels, **kwargs)
+
+
+class TektronixMSO64(TektronixMsoScope):
+    """A class representing the Tektronix MSO64 Scope.
+
+        This scope has 4 analog input and is part of the
+        :class:`~.TektronixMsoScope` family of instruments; for documentation refer to
+        this base class.
+    """
+
+    def __init__(self, adapter,
+                 name="Tektronix MSO64 Oscilloscope",
+                 active_channels=4,
```
If you do set the channels manually here with `active_channels=4` etc, please use the ChannelCreator instead, as described here: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/channels.html#adding-a-channel-with-channelcreator

BenediktBurger reviewed (changes requested):
Thank you for your work and your patience.
